zerver/tests/test_send_email.py

```diff
@@ -57,3 +69,173 @@ def test_build_SES_incompatible_From_field_limit(self) -> None:
             language="en",
         )
         self.assertEqual(mail.extra_headers["From"], FromAddress.NOREPLY)
+
+
+# This method will be used by the mock to replace mail.send
+def mocked_mail_send(self: EmailMultiAlternatives) -> int:
+    if len(sanitize_address(self.extra_headers["From"], "utf-8")) > 320:
+        raise SMTPDataError(code=242, msg="From field is too long.")
+    # Simulate real server response
+    elif not (self.to or self.cc or self.bcc):
+        return 0
+    else:
+        return 1
+
+
+# This method will be used by the mock to replace build_email when the mail
+# should not contain a receiver (to, cc and bcc fields are empty) to trigger
+# a 0 return value (cf. mocked_mail_send)
+def mocked_build_email(
+    template_prefix: str,
+    to_user_ids: Optional[List[int]] = None,
+    to_emails: Optional[List[str]] = None,
+    from_name: Optional[str] = None,
+    from_address: Optional[str] = None,
+    reply_to_email: Optional[str] = None,
+    language: Optional[str] = None,
+    context: Mapping[str, Any] = {},
+    realm: Optional[Realm] = None,
+) -> EmailMultiAlternatives:
+
+    mail = build_email(
+        template_prefix,
+        to_user_ids=to_user_ids,
+        to_emails=to_emails,
+        from_name=from_name,
+        from_address=from_address,
+        reply_to_email=reply_to_email,
+        language=language,
+        context=context,
+        realm=realm,
+    )
+
+    mail.to = None
+    mail.cc = None
+    mail.bcc = None
+    return mail
```

Comment:
This is a lot of test setup code in order to cover this test; could we not just mock `mail.send` to raise the target exception as a `side_effect` with `mock.patch`?  I don't think we want to maintain 150 lines of tests for this tiny bit of exception handler logic.  https://docs.python.org/3/library/unittest.html

Also, I think we don't need the `ok` tests here -- we know that sending emails works from other tests -- so I think we just need to add a single test function to confirm that we do the right thing when this exception is thrown.

(Generally the strategy is to find an existing simple test that already verifies success (potentially as a part of verifying invite emails are sent, or whatever) and then just add an adjacent single test function for the new exception.

Reply:
Yes, it would be more convenient to just use the `side_effect` argument of `mock.patch`, I will change this. I will also delete the ok test as it is redundant with tests in other files. In order to ensure that we do the right thing when an exception is thrown I wanted to test three different cases (represented by the three other tests) : 

- the email can't be sent because no receivers are mentioned (the 0 return value of `mail.send`).
- the address of the sender is longer than 320 bytes and it triggers an `SMTPDataError`.
- an `SMTPException` is triggered (which covers every other possible exception)

Should I keep all of them ? The first one and the third one are essential as they test the two scenarios that would trigger an EmailNotDeliveredException in `send_email` but the second one is a bit redundant with the third one. I wrote it because  I wanted to check that an overly-long From address field indeed triggers an exception but I realize that we already know that right or I wouldn't have implemented the other commit. If you think this second test is not useful I can remove it as well.

Reply:
I think the second is unnecessary, or maybe more correctly, would be better done as replacing one of your existing unit tests for long sender addresses, which are also kinda redundant.

It's also perhaps worth looking at whether we have an existing test that already does the "no receivers" case.